spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java

```diff
@@ -1149,10 +1148,343 @@ public void createViewWithSubqueryExpressionUsingGlobalTempView() {
+    assertThatThrownBy(
+            () -> sql("ALTER VIEW %s AS SELECT non_existing from %s", viewName, tableName))
+        .isInstanceOf(AnalysisException.class)
+        .hasMessageContaining(
+            "A column or function parameter with name `non_existing` cannot be resolved");
+  }
+
+  /** The purpose of this test is mainly to make sure that altering normal view isn't messed up */
+  @Test
+  public void alterViewHiddenByTempView() throws NoSuchTableException {
+    insertRows(6);
+    String viewName = "alteredViewHiddenByTempView";
+    String sql = String.format("SELECT id, data FROM %s WHERE id > 3", tableName);
+
+    sql("CREATE TEMPORARY VIEW %s AS SELECT id FROM %s WHERE id <= 3", viewName, tableName);
+    sql("CREATE VIEW %s AS %s", viewName, sql);
+
+    ViewCatalog viewCatalog = viewCatalog();
+    View view = viewCatalog.loadView(TableIdentifier.of(NAMESPACE, viewName));
+    assertThat(view.sqlFor("spark").sql()).isEqualTo(sql);
+
+    // results from TEMP VIEW are returned
+    assertThat(sql("SELECT id FROM %s", viewName))
+        .hasSize(3)
+        .containsExactlyInAnyOrder(row(1), row(2), row(3));
+
+    // TEMP VIEW will be updated
+    sql("ALTER VIEW %s AS SELECT id FROM %s WHERE id >= 3 AND id < 4", viewName, tableName);
+
+    // results from TEMP VIEW are returned
+    assertThat(sql("SELECT * FROM %s", viewName)).hasSize(1).containsExactly(row(3));
+
+    // the only way to ALTER the Iceberg view would be to rename the TEMP VIEW and then do an ALTER
+    assertThat(view.sqlFor("spark").sql()).isEqualTo(sql);
+  }
+
+  @Test
+  public void alterViewWithCTE() throws NoSuchTableException {
+    insertRows(10);
+    String viewName = viewName("simpleViewWithCTE");
+    String sql =
+        String.format(
+            "WITH max_by_data AS (SELECT max(id) as max FROM %s) "
+                + "SELECT max, count(1) AS count FROM max_by_data GROUP BY max",
+            tableName);
+
+    sql("CREATE VIEW %s AS SELECT * FROM %s", viewName, tableName);
+    sql("ALTER VIEW %s AS %s", viewName, sql);
+
+    assertThat(sql("SELECT * FROM %s", viewName)).hasSize(1).containsExactly(row(10, 1L));
+  }
+
+  @Test
+  public void alterViewWithCTEReferencingTempView() {
+    String viewName = viewName("viewWithCTEReferencingTempView");
+    String tempViewInCTE = viewName("tempViewInCTE");
+    String sql =
+        String.format(
+            "WITH max_by_data AS (SELECT max(id) as max FROM %s) "
+                + "SELECT max, count(1) AS count FROM max_by_data GROUP BY max",
+            tempViewInCTE);
+
+    sql("CREATE TEMPORARY VIEW %s AS SELECT id FROM %s WHERE ID <= 5", tempViewInCTE, tableName);
+    sql("CREATE VIEW %s AS SELECT * FROM %s", viewName, tableName);
+
+    assertThatThrownBy(() -> sql("ALTER VIEW %s AS %s", viewName, sql))
+        .isInstanceOf(AnalysisException.class)
+        .hasMessageContaining("Cannot create the persistent object")
```

Comment:
the error message in Spark unfortunately doesn't distinguish between create/alter cases and throws that error for V1 views as well